cpp/src/lists/apply_boolean_mask.cu

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/fill.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/reduction_functions.hpp>
+#include <cudf/detail/replace.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/lists/detail/stream_compaction.hpp>
+#include <cudf/lists/stream_compaction.hpp>
+#include <cudf/utilities/bit.hpp>
+
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/reduce.h>
+
+namespace cudf::lists {
+namespace detail {
+namespace {
+
+class get_list_size {
+ public:
+  explicit get_list_size(lists_column_view const& lcv)
+    : num_rows{lcv.size()},
+      offsets{lcv.offsets().begin<offset_type>() + lcv.offset()},
+      bitmask{lcv.null_mask()}
+  {
+  }
+
+  size_type __device__ operator()(size_type i) const
+  {
+    return bit_value_or(bitmask, i, true) ? (offsets[i + 1] - offsets[i]) : 0;
+  }
+
+ private:
+  size_type num_rows;
+  offset_type const* offsets;
+  bitmask_type const* bitmask;
+};
+
+void assert_same_list_sizes(lists_column_view const& input,
```

Comment:
I think this assertion will not going to be accepted in cudf. Typically "users are responsible to have the input data valid, any invalid input will produce undef result".

Reply:
There was a similar suggestion on a different PR as well. @jrhemstad's suggestion was to only make the check if an env-variable is set.
It applies here as well, I think.

Reply:
I don't want to abuse the environment variable idea. That was only for very specific use case since it pervaded every interface. For something like this, @ttnghia is correct that we would not introspect the data for this error checking. There are similar APIs the expect list sizes to be identical without checking it.

Reply:
@ttnghia, @jrhemstad, you've convinced me. I'll remove these checks.

Reply:
Safety checks are off. :crossed_fingers: